tests/test-integration/src/test/java/org/eclipse/jetty/test/RebuildSelectorTest.java

```diff
@@ -0,0 +1,254 @@
+            this.strategyFactory = strategyFactory;
+        }
+
+        @Override
+        public ExecutionStrategy.Factory getExecutionStrategyFactory()
+        {
+            return this.strategyFactory;
+        }
+
+        @Override
+        protected SelectorManager newSelectorManager(Executor executor, Scheduler scheduler, int selectors)
+        {
+            return new ServerConnectorManager(executor, scheduler, selectors)
+            {
+                @Override
+                protected ManagedSelector newSelector(int id)
+                {
+                    return new CustomManagedSelector(this, id, getExecutionStrategyFactory());
+                }
+            };
+        }
+    }
+
+    public static class CustomManagedSelector extends ManagedSelector
+    {
+        private static final Logger LOG = Log.getLogger(CustomManagedSelector.class);
+
+        public CustomManagedSelector(SelectorManager selectorManager, int id, ExecutionStrategy.Factory executionFactory)
+        {
+            super(selectorManager, id, executionFactory);
+        }
+
+        @Override
+        protected void onSelectFailed(Throwable cause)
+        {
+            try
+            {
+                LOG.debug("onSelectFailed()", cause);
+                this.startSelector();
+            }
+            catch (Exception ex)
+            {
+                LOG.warn(ex);
+            }
+        }
+    }
+
+    public static class HelloServlet extends HttpServlet
+    {
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
+        {
+            resp.setContentType("text/plain");
+            resp.setCharacterEncoding("utf-8");
+            resp.getWriter().printf("Hello %s:%d%n", req.getRemoteAddr(), req.getRemotePort());
+        }
+    }
+
+    public static class CloseSelectorServlet extends HttpServlet
+    {
+        private static final Logger LOG = Log.getLogger(CloseSelectorServlet.class);
+        private static final int DELAY_MS = 500;
+        private ServerConnector connector;
+        private ScheduledExecutorService scheduledExecutorService;
+        private InterruptSelector interruptSelectorRunnable;
+
+        @Override
+        public void init()
+        {
+            connector = (ServerConnector)getServletContext().getAttribute("connector");
+            scheduledExecutorService = Executors.newScheduledThreadPool(5);
+            interruptSelectorRunnable = new InterruptSelector();
+        }
+
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
+        {
+            resp.setContentType("text/plain");
+            resp.setCharacterEncoding("utf-8");
+            resp.getWriter().printf("Closing selectors in %,d ms%n", DELAY_MS);
+            scheduledExecutorService.schedule(interruptSelectorRunnable, DELAY_MS, TimeUnit.MILLISECONDS);
+        }
```

Comment:
Can we make this a more difficult test (or another test).   Do a startAsync and register a AsyncListener, then return to jetty and have another thread do the interrupt of the selector.    If we are not rescueing end points, then the AsyncListener.onError should be called to tell the async process that you're dead!

Reply:
ew!